### ISSUE TITLE
List parameters in query string format

### ISSUE THREAD

**Issue by LKay** [tiangolo/fastapi#739]
**Description**

Currently, only format with plain parameter name is supported when passing list of parameters https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#query-parameter-list-multiple-values and only `q=1&q=2&q=3` format is supported, however passing list parameters using indexed or not indexed square brackets is widely used across the web and web APIs. None of this is, unfortunately a standard defined in any RFC.

There are 3 most common ways of passing lists out there:
* `q=1&q=2&q=3`
* `q[]=1&q[]=2&q[]=3`
* `q[0]=1&q[1]=2&q[2]=3`

Some sdks, or clients for making requests by default send list using square brackets so there are some issues with configuration quite often.

Are there any plans to support square brackets for list parameters along with the current format?

**Comment by dmontagu** [comment]
Whether this is supported by fastapi would probably be a direct function of whether it was supported by starlette -- fastapi gets the raw query parameter values from starlette.

**Comment by tiangolo** [comment]
Yep. FastAPI gets it directly from Starlette. But still, I would be hesitant of supporting anything like that as, precisely, it's not in any standard and it would be very application-specific. And any client sending a custom format is doing that, sending its own custom format. There's no easy and generalizable way of supporting all possible custom clients.

**Comment by haynesgt** [comment]
This is what I ended up putting in my app:

```
@app.middleware("http")
async def fix_list_query_params(request: fastapi.Request, call_next):
    # Rails / some libs put square brackets on list parameters. FastAPI does not use the square brackets.
    def fix_key(key: str):
        if key and key.endswith("[]"):
            key = key[:-2]
        return key

    request._query_params = QueryParams(
        [(fix_key(key), value) for key, value in request.query_params.multi_items()]
    )
    request.scope["query_string"] = str(request.query_params)
    response = await call_next(request)
    return response
```

**Comment by aimanhey** [comment]
How do you get QueryParams() from? Any libraries?

**Comment by haynesgt** [comment]
`from starlette.datastructures import QueryParams`
